### PR TITLE
fix: possible page_service hang on cancel

### LINKED ISSUE #5341: A pageserver client that's not receiving data blocks timeline deletion
This happened on production a couple of times now:

- timeline deletion is requested
- a compute is connected to the timeline, from pageserver's point of view
- in reality, the compute VM has already been killed, but the pagserver still thinks that the TCP connection is open
- pagserver's task handling the connection is stuck in a write(), because the compute is not receiving the data anymore and the send buffer is full
- as long as the client is connected, the timeline deletion cannot proceed

The situation resolves eventually, when the TCP connection times out, but that can take a very long time. Worse, if the client is still present and simply isn't reading the data the pageserver sends it, it will be stuck forever.

### PR DESCRIPTION
**Pull request by koivunej**
Fixes #5341, one more suspected case, see: https://github.com/neondatabase/neon/issues/5341#issuecomment-1783052379

- races `MaybeWriteOnly::shutdown` with cancellation
- switches to using `AsyncWriteExt::write_buf`
- notes cancellation safety for shutdown

### CODE REVIEW

File: libs/pq_proto/src/framed.rs [hunk 1]
```diff
@@ -214,27 +214,24 @@ where
...
             return Err(io::Error::new(
                 ErrorKind::WriteZero,
                 "failed to write message",
             ));
         }
-        // The advanced part will be garbage collected, likely during shifting
-        // data left on next attempt to write to buffer when free space is not
-        // enough.
-        write_buf.advance(bytes_written);
     }
```
Comment (koivunej) [inline libs/pq_proto/src/framed.rs:230]: Was there a reason why `write_buf` was not used originally? @arssher @petuhovskiy

**Review by jcsp — APPROVED** [review]

File: libs/pq_proto/src/framed.rs (hunk 1 above)
Comment (koivunej) [reply libs/pq_proto/src/framed.rs:230]: Should had not left out the `write_buf.clear()` from the highlight, on the next line.

I think that was extra given how BytesMut reuses the buffer.

Comment (koivunej) [reply libs/pq_proto/src/framed.rs:230]: Toggling auto-merge as I assume there is no reason for not using write_buf as read_buf is used elsewhere. Please do not resolve the conversation too early :)

Comment (petuhovskiy) [reply libs/pq_proto/src/framed.rs:230]: I looked at the BytesMut impl and `write_buf` and `clear` are doing exactly the same, so no reason

**Pull request merged**